File: app/src/main/java/org/dhis2/usescases/teiDashboard/adapters/IndicatorViewHolder.java
package org.dhis2.usescases.teiDashboard.adapters;

import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;

import org.dhis2.BR;
import org.dhis2.R;
import org.dhis2.data.tuples.Trio;
import org.dhis2.databinding.ItemIndicatorBinding;

import org.dhis2.utils.Constants;
import org.dhis2.utils.custom_views.CustomDialog;
import org.hisp.dhis.android.core.program.ProgramIndicatorModel;

/**
 * QUADRAM. Created by ppajuelo on 29/11/2017.
 */

class IndicatorViewHolder extends RecyclerView.ViewHolder {
    ItemIndicatorBinding binding;

    IndicatorViewHolder(ItemIndicatorBinding binding) {
        super(binding.getRoot());
        this.binding = binding;
    }

    public void bind(Trio<ProgramIndicatorModel, String, String> programIndicatorModel) {
        binding.setVariable(BR.indicator, programIndicatorModel.val0());
        binding.setVariable(BR.value, programIndicatorModel.val1());
        binding.setVariable(BR.colorBg, programIndicatorModel.val2().isEmpty() ? -1 : Color.parseColor(programIndicatorModel.val2()));
        binding.executePendingBindings();

        binding.descriptionLabel.setOnClickListener(view->showDescription(programIndicatorModel.val0()));
    }

    private void showDescription(@NonNull ProgramIndicatorModel programIndicatorModel) {
        new CustomDialog(
                itemView.getContext(),
                programIndicatorModel.displayName(),
                programIndicatorModel.displayDescription(),
                itemView.getContext().getString(R.string.action_accept),
                null,
                Constants.DESCRIPTION_DIALOG,
                null
        ).show();
    }
}

File: dhis_android_analytics/src/main/java/dhis2/org/analytics/charts/ui/IndicatorViewHolder.kt
package dhis2.org.analytics.charts.ui

import android.view.ViewGroup
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.material.ripple.rememberRipple
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.unit.dp
import androidx.recyclerview.widget.RecyclerView
import dhis2.org.R
import org.dhis2.commons.dialogs.CustomDialog
import org.hisp.dhis.android.core.program.ProgramIndicator
import org.hisp.dhis.mobile.ui.designsystem.component.IndicatorInput
import org.hisp.dhis.mobile.ui.designsystem.theme.SurfaceColor

class IndicatorViewHolder(
    val parent: ViewGroup,
) : RecyclerView.ViewHolder(ComposeView(parent.context)) {

    fun bind(programIndicatorModel: IndicatorModel) {
        val view = itemView as ComposeView
        view.apply {
            layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT,
            )

            setContent {
                Column(
                    Modifier
                        .fillMaxSize()
                        .then(
                            if (programIndicatorModel.programIndicator?.description() != null) {
                                Modifier.clickable(
                                    interactionSource = MutableInteractionSource(),
                                    indication = rememberRipple(),
                                    onClick = { showDescription(programIndicatorModel.programIndicator) },
                                )
                            } else {
                                Modifier
                            },
                        ),
                ) {
                    IndicatorInput(
                        title = programIndicatorModel.label(),
                        indicatorColor = if (!programIndicatorModel.color.isNullOrEmpty()) {
                            Color(programIndicatorModel.color())
                        } else {
                            SurfaceColor.Container
                        },
                        content = programIndicatorModel.value ?: "",
                        modifier = Modifier.then(
                            if (programIndicatorModel.programIndicator?.description() != null) {
                                Modifier.clickable(
                                    interactionSource = MutableInteractionSource(),
                                    indication = rememberRipple(),
                                    onClick = { showDescription(programIndicatorModel.programIndicator) },
                                )
                            } else {
                                Modifier
                            },
                        ),
                    )
                    Spacer(modifier = Modifier.size(16.dp))
                }
            }
        }
    }

    private fun showDescription(programIndicatorModel: ProgramIndicator?) {
        programIndicatorModel?.let {
            CustomDialog(
                itemView.context,
                it.displayName() ?: "",
                it.displayDescription() ?: "",
                itemView.context.getString(R.string.action_accept),
                null,
                CustomDialog.DESCRIPTION_DIALOG,
                null,
            ).show()
        }
    }
}
